subsys/net/l2/ethernet/bridge.c

```diff
@@ -10,144 +11,198 @@ LOG_MODULE_REGISTER(net_eth_bridge, CONFIG_NET_ETHERNET_BRIDGE_LOG_LEVEL);
-	 * macro assumed that k_mutex can be statically initialized,
-	 * and it can't.  Post-zync, this will actually be possible
-	 * and we can come back and fix this.
-	 */
-	if (!br->initialized) {
-		k_mutex_init(&br->lock);
-		br->initialized = true;
-	}
-	k_mutex_lock(&br->lock, K_FOREVER);
+	k_mutex_lock(&ctx->lock, K_FOREVER);
 }
 
-void net_eth_bridge_foreach(eth_bridge_cb_t cb, void *user_data)
+static void unlock_bridge(struct eth_bridge_iface_context *ctx)
 {
-	STRUCT_SECTION_FOREACH(eth_bridge, br) {
-		cb(br, user_data);
-	}
+	k_mutex_unlock(&ctx->lock);
 }
 
-int eth_bridge_get_index(struct eth_bridge *br)
+struct ud {
+	eth_bridge_cb_t cb;
+	void *user_data;
+};
+
+static void iface_cb(struct net_if *iface, void *user_data)
 {
-	if (!(br >= _eth_bridge_list_start && br < _eth_bridge_list_end)) {
-		return -1;
+	struct ud *br_user_data = user_data;
+	struct eth_bridge_iface_context *ctx;
+	enum virtual_interface_caps caps;
+
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(VIRTUAL)) {
+		return;
+	}
+
+	caps = net_virtual_get_iface_capabilities(iface);
+	if (!(caps & VIRTUAL_INTERFACE_BRIDGE)) {
+		return;
 	}
 
-	return (br - _eth_bridge_list_start) + 1;
+	ctx = net_if_get_device(iface)->data;
+
+	br_user_data->cb(ctx, br_user_data->user_data);
 }
 
-struct eth_bridge *eth_bridge_get_by_index(int index)
+void net_eth_bridge_foreach(eth_bridge_cb_t cb, void *user_data)
 {
-	if (index <= 0) {
-		return NULL;
-	}
+	struct ud br_user_data = {
+		.cb = cb,
+		.user_data = user_data,
+	};
 
-	if (&_eth_bridge_list_start[index - 1] >= _eth_bridge_list_end) {
-		NET_DBG("Index %d is too large", index);
-		return NULL;
-	}
+	net_if_foreach(iface_cb, &br_user_data);
+}
+
+int eth_bridge_get_index(struct net_if *br)
+{
+	return net_if_get_by_iface(br);
+}
 
-	return &_eth_bridge_list_start[index - 1];
+struct net_if *eth_bridge_get_by_index(int index)
+{
+	return net_if_get_by_index(index);
 }
 
-int eth_bridge_iface_add(struct eth_bridge *br, struct net_if *iface)
+int eth_bridge_iface_add(struct net_if *br, struct net_if *iface)
 {
-	struct ethernet_context *ctx = net_if_l2_data(iface);
+	struct eth_bridge_iface_context *ctx = net_if_get_device(br)->data;
+	struct ethernet_context *eth_ctx = net_if_l2_data(iface);
+	bool found = false;
+	int count = 0;
+	int ret;
 
 	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET) ||
 	    !(net_eth_get_hw_capabilities(iface) & ETHERNET_PROMISC_MODE)) {
 		return -EINVAL;
 	}
 
-	lock_bridge(br);
-
-	if (ctx->bridge.instance != NULL) {
-		k_mutex_unlock(&br->lock);
-		return -EBUSY;
+	if (net_if_l2(br) != &NET_L2_GET_NAME(VIRTUAL) ||
+	    !(net_virtual_get_iface_capabilities(br) & VIRTUAL_INTERFACE_BRIDGE)) {
+		return -EINVAL;
 	}
 
-	ctx->bridge.instance = br;
-	ctx->bridge.allow_tx = false;
-	sys_slist_append(&br->interfaces, &ctx->bridge.node);
+	lock_bridge(ctx);
+
+	ARRAY_FOR_EACH(ctx->eth_iface, i) {
+		if (!found && ctx->eth_iface[i] == NULL) {
+			ctx->eth_iface[i] = iface;
+			eth_ctx->bridge = br;
```

Comment:
In this case the `eth_ctx->bridge` points to the actual virtual bridge interface, it is not the bridged Ethernet interface.
But indeed, we should check that we are not adding the bridged Ethernet interface twice to the list.

Reply:
It looks we can just do this
```
	if (eth_ctx->bridge == br) {
		/* This Ethernet interface was already added to the bridge */
		found = true;
	}
```
before the `ARRAY_FOR_EACH` loop, so that the bridge interface is not added twice.